src/rust/src/rustzcash.rs

```diff
@@ -449,34 +455,40 @@ pub extern "C" fn librustzcash_sapling_compute_cmu(
     true
 }
 
-/// Computes \[sk\] \[8\] P for some 32-byte point P, and 32-byte Fs.
+/// Computes KDF^Sapling(KA^Agree(sk, P), ephemeral_key).
 ///
-/// If P or sk are invalid, returns false. Otherwise, the result is written to
-/// the 32-byte `result` buffer.
+/// `p` and `sk` must point to 32-byte buffers. If `p` does not represent a compressed
+/// Jubjub point or `sk` does not represent a canonical Jubjub scalar, this function
+/// returns `false`. Otherwise, it writes the result to the 32-byte `result` buffer and
+/// returns `true`.
 #[no_mangle]
-pub extern "C" fn librustzcash_sapling_ka_agree(
+pub extern "C" fn librustzcash_sapling_ka_derive_symmetric_key(
     p: *const [c_uchar; 32],
     sk: *const [c_uchar; 32],
+    ephemeral_key: *const [c_uchar; 32],
     result: *mut [c_uchar; 32],
 ) -> bool {
-    // Deserialize p
-    let p = match de_ct(jubjub::ExtendedPoint::from_bytes(unsafe { &*p })) {
-        Some(p) => p,
+    // Deserialize p (representing either epk or pk_d; we can handle them identically).
+    let epk = match SaplingDomain::<Network>::epk(&EphemeralKeyBytes(unsafe { *p })) {
+        Some(p) => SaplingDomain::<Network>::prepare_epk(p),
         None => return false,
     };
 
-    // Deserialize sk
-    let sk = match de_ct(jubjub::Scalar::from_bytes(unsafe { &*sk })) {
-        Some(p) => p,
+    // Deserialize sk (either ivk or esk; we can handle them identically).
+    let ivk = match de_ct(jubjub::Scalar::from_bytes(unsafe { &*sk })) {
+        Some(p) => PreparedIncomingViewingKey::new(&SaplingIvk(p)),
         None => return false,
     };
 
     // Compute key agreement
-    let ka = sapling_ka_agree(&sk, &p);
+    let secret = SaplingDomain::<Network>::ka_agree_dec(&ivk, &epk);
```

Comment:
This uses [PreparedEphemeralPublicKey::agree(PreparedIncomingViewingKey)](https://github.com/zcash/librustzcash/blob/zcash_primitives-0.10.0/zcash_primitives/src/sapling/note_encryption.rs#L172-L177) (because [SaplingDomain::IncomingViewingKey = PreparedIncomingViewingKey](https://github.com/zcash/librustzcash/blob/zcash_primitives-0.10.0/zcash_primitives/src/sapling/note_encryption.rs#L139)) which ends up using [ka_sapling_agree_prepared](https://github.com/zcash/librustzcash/blob/zcash_primitives-0.10.0/zcash_primitives/src/sapling/spec.rs#L124-L136), so I agree (no pun intended) that it's doing the same thing.